src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportJobTask.cs

```diff
@@ -218,7 +167,131 @@ private async Task UpdateJobRecordAsync(CancellationToken cancellationToken)
+                    {
+                        // If a job is resumed in the middle of processing patient compartment resources it will skip patients it has already exported compartment information for.
+                        if (progress.SubSearch != null && result.Resource.ResourceId != progress.SubSearch.TriggeringResourceId)
+                        {
+                            continue;
+                        }
+
+                        if (progress.SubSearch == null)
+                        {
+                            progress.NewSubSearch(result.Resource.ResourceId);
+                        }
+
+                        await RunExportSearch(exportJobConfiguration, progress.SubSearch, sharedQueryParametersList, cancellationToken, pagesWaitingForCommit, currentBatchId + ":" + resultIndex.ToString("d6"), true);
+                        resultIndex++;
+
+                        progress.ClearSubSearch();
+                    }
+                }
+
+                await ProcessSearchResultsAsync(searchResult.Results, currentBatchId, cancellationToken);
+
+                if (searchResult.ContinuationToken == null)
+                {
+                    // No more continuation token, we are done.
+                    break;
+                }
+
+                // Update the continuation token in local cache and queryParams.
+                // We will add or udpate the continuation token to the end of the query parameters list.
+                progress.UpdateContinuationToken(searchResult.ContinuationToken);
+                pagesWaitingForCommit++;
+
+                if (queryParametersList[queryParametersList.Count - 1].Item1 == KnownQueryParameterNames.ContinuationToken)
+                {
+                    queryParametersList[queryParametersList.Count - 1] = Tuple.Create(KnownQueryParameterNames.ContinuationToken, progress.ContinuationToken);
+                }
+                else
+                {
+                    queryParametersList.Add(Tuple.Create(KnownQueryParameterNames.ContinuationToken, progress.ContinuationToken));
+                }
+
+                if (progress.Page % _exportJobRecord.NumberOfPagesPerCommit == 0 || (_exportJobRecord.ExportType == ExportJobType.Patient && !compartmentSearch))
+                {
+                    // Commit the changes.
+                    await _exportDestinationClient.CommitAsync(exportJobConfiguration, cancellationToken);
+
+                    // Update the job record.
+                    await UpdateJobRecordAsync(cancellationToken);
+
+                    currentBatchId = batchIdPrefix + '-' + progress.Page.ToString("d6");
+                    pagesWaitingForCommit = 0;
+                }
+            }
+
+            // Commit one last time for any pending changes.
+            await _exportDestinationClient.CommitAsync(exportJobConfiguration, cancellationToken);
+            await UpdateJobRecordAsync(cancellationToken);
```

Comment:
After the method exits, we call `CompleteJobAsync()` which calls `UpdateJobRecordAsync()`. We can avoid the `UpdateJobRecordAsync()` call here I think

Reply:
I added this to account for the recursive nature of the method. Since there can now be sub searches this records the progress when a sub search ends.